File: day-01/python_jackr/day01.py
import re

with open("input.txt") as f:
    lines = f.readlines()

# PART 1
total = 0
for line in lines:
    digits_str = re.findall("[0-9]", line)
    first_last = int("".join([digits_str[0], digits_str[-1]]))
    total += first_last

print("Part 1", total)

# PART 2
patterns = [
    "one",
    "1",
    "two",
    "2",
    "three",
    "3",
    "four",
    "4",
    "five",
    "5",
    "six",
    "6",
    "seven",
    "7",
    "eight",
    "8",
    "nine",
    "9",
]
word_to_digit = {
    "one": "1",
    "two": "2",
    "three": "3",
    "four": "4",
    "five": "5",
    "six": "6",
    "seven": "7",
    "eight": "8",
    "nine": "9",
}


def pattern_in_str(string):
    for p in patterns:
        if p in string:
            if p in word_to_digit:
                return word_to_digit[p]
            return p
    return ""


total = 0
for line in lines:
    idx = 0
    first = ""
    while not first:
        first = pattern_in_str(line[:idx])
        idx += 1

    idx = -1
    last = ""
    while not last:
        last = pattern_in_str(line[idx:])
        idx -= 1
    first_last = int("".join([first, last]))
    total += first_last

print("Part 2", total)

File: day-01/r_phil/day01.R
# Read in file ----
file <- "data.txt"
vec <- scan(file, character(), sep="\n", quote=NULL)

#  Fns ----
is_digit <- function(character) !is.na(suppressWarnings(as.numeric(character)))

find_digit <- function(string, n, operator) {
  character <- substring(string, n, n)
  if (is_digit(character)) return(character)
  else return(find_digit(string, operator(n, 1), operator))
}

words <- c("one", "two", "three", "four", "five",
           "six", "seven", "eight", "nine")
numbers <- 1:9

apply_replacements <- function(string, replacements, num=1) {
  indices <- replacements[[num]]
  for (i in indices) substr(string, i, i) <- as.character(num)
  if (num == 9) return(string)
  return(apply_replacements(string, replacements, num+1))
}

make_replacements <- function(string) {
  replacements <- lapply(numbers, function(x) gregexpr(words[x], string)[[1]])
  return(apply_replacements(string, replacements))
}

sum_values <- function(vec, replace) {
  if (replace) vec <- sapply(vec, function(x) make_replacements(x))
  first <- sapply(vec, function(x) find_digit(x, 1, `+`))
  last <- sapply(vec, function(x) find_digit(x, nchar(x), `-`))
  return(sum(as.numeric(paste0(first, last))))
}

# Part 1 ----
sum_values(vec, F)

# Part 2 ----
sum_values(vec, T)

